<a href="https://colab.research.google.com/github/LiteraryVectors/Python_library/blob/Data_Science_Techniques/Ghost_Vectors.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Test Project (Just for Fun)
# The Subjectivity of Gendered Ghosts in Mattew Lewis' *The Monk* and Horace Walpole's *The Castle of Otranto*

![alt text](https://upload.wikimedia.org/wikipedia/commons/b/b6/Ghost_Stories_May_1931.jpg)

In [1]:
!python -m pip install pysimplegui


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


# Part One: 

The first section of this exam will include an attempt to explore my reading of two Gothic novels utilizing functions from my own repository, slicing strings to cut the Gutenburg.org headers, creating lists, writing *for* loops to update a data frame, sorting a word bag by author name and storing to the results to an outside file which can be read back in, and analyzing the word instances from a humanist prospective. 

The two novels explored below are some of the earliest examples in the genre of Gothic Literature, Horace Walpole's *The Castle of Otranto*, and Matthew Lewis' *The Monk* (both 18th century). They are often read as artifacts of several social tensions, including gender representation. Specifically, these novels were highly sexist.

**Research Questions: Are women talked about as objects in the way ghosts are? Do women hold more subjectivity in these strikingly androcentric novels than that of ghosts?**


In [2]:
import pandas as pd

In [3]:
!ls

601-0.txt    601-0.txt.3  696-0.txt    696-0.txt.3  Python_library
601-0.txt.1  601-0.txt.4  696-0.txt.1  696-0.txt.4
601-0.txt.2  601-0.txt.5  696-0.txt.2  gdrive


In [4]:
!rm -r Python_library	sample_data
LiteraryVectors = 'LiteraryVectors'
clone_url = f'https://github.com/{LiteraryVectors}/Python_library.git'
!git clone $clone_url


rm: cannot remove 'sample_data': No such file or directory
Cloning into 'Python_library'...
remote: Enumerating objects: 240, done.
remote: Counting objects: 100% (216/216), done.
remote: Compressing objects: 100% (114/114), done.
remote: Total 240 (delta 122), reused 173 (delta 99), pack-reused 24
Receiving objects: 100% (240/240), 79.29 KiB | 4.40 MiB/s, done.
Resolving deltas: 100% (124/124), done.


In [5]:
import spacy
!python -m spacy download en_core_web_lg

2022-08-20 18:02:33.539516: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 587.7 MB 7.6 kB/s 
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_lg')


In [6]:

from spacy.lang.en.examples import sentences 
import en_core_web_lg
nlp = spacy.load("en_core_web_lg")

In [7]:
import Python_library.Bayes_Theorem_Pandas_Word_Bag_Heat_Map as pl

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [8]:
Otranto_url = 'http://www.gutenberg.org/files/696/696-0.txt'
Monk_url = 'https://www.gutenberg.org/files/601/601-0.txt'

In [9]:
!wget {Otranto_url}

--2022-08-20 18:03:27--  http://www.gutenberg.org/files/696/696-0.txt
Resolving www.gutenberg.org (www.gutenberg.org)... 152.19.134.47, 2610:28:3090:3000:0:bad:cafe:47
Connecting to www.gutenberg.org (www.gutenberg.org)|152.19.134.47|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://www.gutenberg.org/files/696/696-0.txt [following]
--2022-08-20 18:03:27--  https://www.gutenberg.org/files/696/696-0.txt
Connecting to www.gutenberg.org (www.gutenberg.org)|152.19.134.47|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 240149 (235K) [text/plain]
Saving to: ‘696-0.txt.5’

696-0.txt.5         100%[===================>] 234.52K  --.-KB/s    in 0.09s   

2022-08-20 18:03:27 (2.47 MB/s) - ‘696-0.txt.5’ saved [240149/240149]



In [10]:
!ls -l

total 6296
-rw-r--r-- 1 root root 828670 Mar  8  2021 601-0.txt
-rw-r--r-- 1 root root 828670 Mar  8  2021 601-0.txt.1
-rw-r--r-- 1 root root 828670 Mar  8  2021 601-0.txt.2
-rw-r--r-- 1 root root 828670 Mar  8  2021 601-0.txt.3
-rw-r--r-- 1 root root 828670 Mar  8  2021 601-0.txt.4
-rw-r--r-- 1 root root 828670 Mar  8  2021 601-0.txt.5
-rw-r--r-- 1 root root 240149 Apr  9  2021 696-0.txt
-rw-r--r-- 1 root root 240149 Apr  9  2021 696-0.txt.1
-rw-r--r-- 1 root root 240149 Apr  9  2021 696-0.txt.2
-rw-r--r-- 1 root root 240149 Apr  9  2021 696-0.txt.3
-rw-r--r-- 1 root root 240149 Apr  9  2021 696-0.txt.4
-rw-r--r-- 1 root root 240149 Apr  9  2021 696-0.txt.5
drwx------ 5 root root   4096 Aug 20 15:29 gdrive
drwxr-xr-x 4 root root   4096 Aug 20 18:03 Python_library


In [11]:
with open('696-0.txt', 'r') as f:
  otranto = f.read()

In [12]:
otranto.index('SONNET')

12178

In [13]:
otranto = otranto[6633:]

In [14]:
otranto[0:500]

'think that the date of the composition was little antecedent\nto that of the impression. Letters were then in their most flourishing\nstate in Italy, and contributed to dispel the empire of superstition, at\nthat time so forcibly attacked by the reformers. It is not unlikely that\nan artful priest might endeavour to turn their own arms on the\ninnovators, and might avail himself of his abilities as an author to\nconfirm the populace in their ancient errors and superstitions. If this\nwas his view, he h'

In [15]:
!wget {Monk_url}

--2022-08-20 18:03:28--  https://www.gutenberg.org/files/601/601-0.txt
Resolving www.gutenberg.org (www.gutenberg.org)... 152.19.134.47, 2610:28:3090:3000:0:bad:cafe:47
Connecting to www.gutenberg.org (www.gutenberg.org)|152.19.134.47|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 828670 (809K) [text/plain]
Saving to: ‘601-0.txt.6’

601-0.txt.6         100%[===================>] 809.25K  --.-KB/s    in 0.1s    

2022-08-20 18:03:28 (6.72 MB/s) - ‘601-0.txt.6’ saved [828670/828670]



In [16]:
!ls -l

total 7108
-rw-r--r-- 1 root root 828670 Mar  8  2021 601-0.txt
-rw-r--r-- 1 root root 828670 Mar  8  2021 601-0.txt.1
-rw-r--r-- 1 root root 828670 Mar  8  2021 601-0.txt.2
-rw-r--r-- 1 root root 828670 Mar  8  2021 601-0.txt.3
-rw-r--r-- 1 root root 828670 Mar  8  2021 601-0.txt.4
-rw-r--r-- 1 root root 828670 Mar  8  2021 601-0.txt.5
-rw-r--r-- 1 root root 828670 Mar  8  2021 601-0.txt.6
-rw-r--r-- 1 root root 240149 Apr  9  2021 696-0.txt
-rw-r--r-- 1 root root 240149 Apr  9  2021 696-0.txt.1
-rw-r--r-- 1 root root 240149 Apr  9  2021 696-0.txt.2
-rw-r--r-- 1 root root 240149 Apr  9  2021 696-0.txt.3
-rw-r--r-- 1 root root 240149 Apr  9  2021 696-0.txt.4
-rw-r--r-- 1 root root 240149 Apr  9  2021 696-0.txt.5
drwx------ 5 root root   4096 Aug 20 15:29 gdrive
drwxr-xr-x 4 root root   4096 Aug 20 18:03 Python_library


In [17]:
with open('601-0.txt', 'r') as f:
  monk = f.read()

In [18]:
monk.index('Methinks')

1268

In [19]:
monk = monk[1268:] #attempt to slice off the intro material (it is best to also slice off the Gutenberg.org endnotes, but I do not take that step here)

In [20]:
monk[0:150] # this should show us where we are starting in the text

'Methinks, Oh! vain ill-judging Book,\nI see thee cast a wishful look,\nWhere reputations won and lost are\nIn famous row called Paternoster.\nIncensed to '

In [21]:
doc1 = nlp(otranto.lower())

In [22]:
doc2 = nlp(monk.lower())

In [23]:
word_table = pd.DataFrame(columns=['word', 'ML', 'HW'])

In [24]:
for token in doc1:
    if token.is_alpha and not token.is_stop:
      pl.update_gothic_row(word_table,token.text,'HW')

In [25]:
for token in doc2:
    if token.is_alpha and not token.is_stop:
      pl.update_gothic_row(word_table,token.text,'ML')

In [26]:
word_table.head()

word  ML  HW
0        think  80  30
1         date   2   6
2  composition   2   1
3       little  99  33
4   antecedent   0   1

In [27]:
sorted_word_table = word_table.sort_values(by=['word'])
sorted_word_table = sorted_word_table.reset_index(drop=True)

In [28]:
sorted_word_table = sorted_word_table.set_index('word')
sorted_word_table.head()

ML HW
word             
abandon     23  2
abandoned   32  2
abandoning   3  0
abated       7  0
abbess       2  0

In [29]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [30]:
with open('gdrive/My Drive/FinalWordBag.csv', 'w') as file:
    sorted_word_table.to_csv(file, index=True)

In [31]:
with open('gdrive/My Drive/FinalWordBag.csv', 'r') as file:
    sorted_word_table = pd.read_csv(file, dtype={'word':str}, encoding='utf-8',
                                    index_col='word', na_filter=False)

In [32]:
sorted_word_table.loc['women', 'ML'] 

23

In [33]:
sorted_word_table[3746:3750]

ML  HW
word           
ghastly   1   1
ghost    51   7
ghostly   2   0
ghosts    7   1

* ML:53 instances of ghost or ghostly
* HW: 7 instances of ghost or ghostly

In [34]:
sorted_word_table[9442:9446]

ML  HW
word            
woman     91  16
womanish   0   2
womanly    0   1
women     23   7

* ML: 114 instances of the word woman and its variations
* HW: 26 instances of the word woman and its variations 

* ML: 134 instances of the word man and its variations compared to the 114 for women
* HW: 87 instances of the word man and its variations compared to the 26


* ML:52 instances of ghost or ghostly 
* HW: 7 instances of ghost or ghostly


In Horace Walpole's novel *The Castle of Otranto*, there are only 19 more instances of women than ghosts, and in Mathew Lewis' novel *The Monk*, there are 62 more instances of the subject of women over that of ghosts. 

But what does that mean? 
Maybe nothing. 
Let's explore further with word-embedding and sentence vectors. **I am curious if ghosts and women are talked about in similar ways (as an objectified other)**  

# Part Two: 

Vector functions and matrix loops in a word-embedding algorithm to further explore the novels' words of interest within the world of sentences. 

In [35]:
otranto_sentences = list(doc1.sents)

In [36]:
monk_sentences = list(doc2.sents)

In [37]:
sent_table = pd.DataFrame(columns=['Sent', 'ML', 'HW'])

In [38]:
#for i in range (len(otranto_sentences)):
    #pl.update_gothic_row_sent(sent_table,doc1.sents,'HW')
#Experimental cell

In [39]:
#for i in range (len(monk_sentences)):
    #pl.update_gothic_row_sent(sent_table,doc2.sents,'ML')
#Experimental cell

In [40]:
#sent_table.head()
#Experimental cell

In [41]:
otranto_matrix = []

for i in range(len(otranto_sentences)): 
  sentence = otranto_sentences[i]
  vec = pl.sent2vec(sentence.text)
  otranto_matrix.append(vec)

In [42]:
nlp.vocab.has_vector('barrier') #you can check to see if there is data to work with on specific words

True

In [43]:
monk_matrix = []

for i in range(len(monk_sentences)):  
  sentence = monk_sentences[i]
  vec = pl.sent2vec(sentence.text)
  monk_matrix.append(vec)

In [44]:
test_sentence = "ghosts and women haunt and watch"

In [45]:
input_vec = pl.sent2vec(test_sentence)

In [46]:
nlp.vocab.has_vector('subjects')

True

In [47]:
import numpy as np

In [48]:
ordered_distances_otranto = []

for i in range(len(otranto_matrix)):
  vec = otranto_matrix[i]
  d = pl.fast_cosine(np.array(input_vec), np.array(vec))
  ordered_distances_otranto.append([d, i])

In [49]:
for d,j in sorted(ordered_distances_otranto, reverse=True)[:10]:
  print(otranto_sentences[j])
  print('~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~')


he does not
know you, but takes you for one of my lady hippolita’s women.”
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
but there is my lady isabella would not be so reserved to me: she will
let me talk to her of young men: and when a handsome cavalier has come to
the castle, she has owned to me that she wished your brother conrad
resembled him.”
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
come, be a good girl and tell me; is there
any young man—ha!—you understand me.”
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
not one of us your highness’s faithful servants—indeed
we are, my lord, though poor men—i say, not one of us has dared to set a
foot about the castle, but two together: so diego and i, thinking that my
young lady might be in the great gallery, went up there to look for her,
and tell her your highness wanted something to impart to her.”
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
what could a friendless, helpless woman do?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
dear lady, i would not speak to a ghost for the 

In [50]:
ordered_distances_monk = []

for i in range(len(monk_matrix)):
  vec = monk_matrix[i]
  d = pl.fast_cosine(np.array(input_vec), np.array(vec))
  ordered_distances_monk.append([d, i])

In [51]:
for d,j in sorted(ordered_distances_monk, reverse=True)[:10]:
  print(monk_sentences[j])
  print('~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~')

are you not infamous in
the eyes of men and angels.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
then will come your mother’s ghost,
and hurl me down into the dwellings of fiends, and flames, and furies,
and everlasting torments!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
dare you
spring without fear over the bounds which separate men from angels?—you
are silent.—you look upon me with eyes of suspicion and alarm—i read
your thoughts and confess their justice.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
i am
the most unfortunate woman alive!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
amidst these
dismal scenes wandered the ghosts of elvira and her daughter.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
throw from you these terrors so
ill-suited to a soul like yours; leave them for common men, and dare to
be happy!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
you should
not seem to remember that there is such a thing as a man in the world,
and you ought to imagine every body to be of the same sex with
yourself.
~~~~~~~~~~~~~~~~~~~

In [68]:
test_sentence2 = "women and knowing"

In [69]:
input_vec = pl.sent2vec(test_sentence2)

In [70]:
nlp.vocab.has_vector('ways')

True

In [71]:
ordered_distances_otranto = []

for i in range(len(otranto_matrix)):
  vec = otranto_matrix[i]
  d = pl.fast_cosine(np.array(input_vec), np.array(vec))
  ordered_distances_otranto.append([d, i])

In [72]:
for d,j in sorted(ordered_distances_otranto, reverse=True)[:10]:
  print(otranto_sentences[j])
  print('~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~')

he does not
know you, but takes you for one of my lady hippolita’s women.”
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
“but can i, most respected of women, can i experience all
this tenderness, this world of goodness, and conceal a thought from the
best of mothers?”

“what art thou going to utter?” said isabella trembling.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
“my councils do not need a friar’s intervention,” said manfred; “and of
all men living is that hoary traitor the only one whom you delight to
confer with?”

“profane prince!”
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
the company had no sooner quitted the castle than matilda, who felt
herself deeply interested for the young peasant, since she had seen him
condemned to death in the hall, and whose thoughts had been taken up with
concerting measures to save him, was informed by some of the female
attendants that manfred had despatched all his men various ways in
pursuit of isabella.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
“thou and
i must never hop

In [57]:
ordered_distances_monk = []

for i in range(len(monk_matrix)):
  vec = monk_matrix[i]
  d = pl.fast_cosine(np.array(input_vec), np.array(vec))
  ordered_distances_monk.append([d, i])

In [58]:
for d,j in sorted(ordered_distances_monk, reverse=True)[:10]:
  print(monk_sentences[j])
  print('~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~')

believe me, unhappy woman!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
know, vain man!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
foolish woman!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
i am
the most unfortunate woman alive!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
unfeeling woman!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
miserable woman that i
am!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
a man! speak!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
well, to be sure, never was woman so unfortunate, as
myself!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
not know your friend?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
abandoned woman, speak for him no
more!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


For me, test_sentence2 returns passages with significant andro-centric gendering. Also, there is a notable difference here in the style of sentences returned for each novel. The Monk (this latter set above) would appear to be full of short sentences, but many of these lines are truncated; whereas, the results for Otranto include full, and multiple lines of dialog. 


These tests were inspired by the following passage, found here(https://www.gutenberg.org/files/601/601-0.txt):

"“Holy St. Barbara!” exclaimed Leonella; “What a question! Fye, Child,
Fye! These are not fit subjects for young Women to handle. You should
not seem to remember that there is such a thing as a Man in the world,
and you ought to imagine every body to be of the same sex with
yourself. I should like to see you give people to understand, that you
know that a Man has no breasts, and no hips, and no ...”.

Luckily for Antonia’s ignorance which her Aunt’s lecture would soon
have dispelled, an universal murmur through the Church announced the
Preacher’s arrival. Donna Leonella rose from her seat to take a better
view of him, and Antonia followed her example."

It is important to note that my thinking syncing up with some of these results could be just a coincidence, or "the existence of spectres."

**This experiment is full of human bias starting with the fact that the word-embedding algorithm pulls "meaning" from the internet and uses a logic system that has been showed to hold gender bias (doctor-man=nurse). However, there is also a bias on the part of the user. I had close read each novel in advance of these tests and used what I concluded to direct both my search of the word bag and my creation of each test sentence.**